### PR TITLE
feat: underlying typed graphql wrapper

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

Adds types for V6 `graphql()` function, which use correct types for:

1. Branded query strings in the form of `"graphql ... " as GeneratedQuery<IN, OUT>` per the POC.
2. Type argument in the form of `graphql<{variables: VARS_TYPE, result: RESULT_TYPE}>({query: "..."})`;
3. Type argument in the form of `graphql<RESULT_TYPE>({query: "..."})` (Already supported in V5 today.)

#### Description of how you validated changes

1. Added code-generated schemas/graphql for both branded and un-branded graphql
2. Added tests with explicit typing expectations for each ^^^
3. Enabled jest diagnostics -- if return types from `graphql()` break, tests will fail

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api-graphql/__tests__/utils/expects.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+//  * is of the given type `T`.
+//  *
+//  * This will succesfully cause a build to fail for non-T arguments when jest
+//  * diagnostics are enabled.
+//  *
+//  * @param item The item to type-test.
+//  */
+// export function expectType<T>(item: T) {
+// 	return true;
+// }
```
Comment (svidgen) [inline packages/api-graphql/__tests__/utils/expects.ts:121]: Cruft. Remove.

File: packages/api-graphql/src/internals/v6.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	 * Some consideration was given to using a type guard on `result` before returning it.
+	 * And, with some graphql parsing, a type guard could validate that `result` matches the
+	 * shape dictated by `options`. But, `options.query` is only "branded" as `TYPED_GRAPHQL_STRING`.
+	 * Validating that `options` and `result` match *each other* does not prove anything we don't
+	 * already know &mdash; or that `result` is an instance of
+	 * `GraphQLResponseV6<FALLBACK_TYPES, TYPED_GQL_STRING>`.
+	 *
+	 * Per the note at the top, the correctness of these typings depends on the correct
+	 * branding or type overrides, neither of which can actually be validated at runtime.
+	 */
```
Comment (svidgen) [inline packages/api-graphql/src/internals/v6.ts:39]: Too verbose. Shorten.

File: packages/api-graphql/src/types/index.ts [hunk 3]
```diff
@@ -4,8 +4,12 @@ import { Source, DocumentNode, GraphQLError, OperationTypeNode } from 'graphql';
 export { OperationTypeNode } from 'graphql';
 import { GRAPHQL_AUTH_MODE } from '@aws-amplify/auth';
 export { GRAPHQL_AUTH_MODE };
-import { CustomUserAgentDetails } from '@aws-amplify/core';
+import { Observable } from 'zen-observable-ts';
+// import { CustomUserAgentDetails } from '@aws-amplify/core';
```
Comment (svidgen) [inline packages/api-graphql/src/types/index.ts:8]: Remove.

File: packages/api-graphql/src/types/index.ts [hunk 4]
```diff
@@ -17,6 +21,9 @@ export interface GraphQLOptions {
 	userAgentSuffix?: string; // TODO: remove in v6
 }
 
+/**
+ *
```
Comment (svidgen) [inline packages/api-graphql/src/types/index.ts:25]: Complete.

File: packages/api-graphql/src/types/index.ts [hunk 5]
```diff
@@ -25,6 +28,51 @@ export interface GraphQLResult<T = object> {
...
+ * 	//            |-- You are here
+ * 	//            v
+ * 	next(message: GraphqlSubscriptionMessage<OnCreateWidgetSubscription>) {
+ * 		handle(message.value);  // <-- type OnCreateWidgetSubscription
+ * 	}
+ * })
+ * ```
+ */
+export type GraphqlSubscriptionMessage<T> = {
+	provider: any;
```
Comment (iartemiev) [inline packages/api-graphql/src/types/index.ts:72]: This is the configured PubSub provider, right? Any reason we can't type this?

**Review by iartemiev — APPROVED** [review]

File: packages/api-graphql/src/types/index.ts (hunk 5 above)
Comment (svidgen) [reply packages/api-graphql/src/types/index.ts:72]: As discussed, I'll follow up on this in another PR today.

**Pull request merged**